### DISCUSSION TITLE
Middleware _MiddlewareClass mypy typing PR2381

### DISCUSSION THREAD

**Discussion by ShedPlant** [encode/starlette#2581] [Potential Issue] [answered]
I have some custom middleware a bit like

```python
class TraceTokenLoggerMiddleware:
    def __init__(self, app, logger: IAsyncLogger):
        self._app = app
        self._async_logger = logger

    async def __call__(self, scope: dict, receive, send):
         # do something
        pass
```

Then creating my app:
```python
App(
    ...
    middleware=[
            Middleware(TraceTokenLoggerMiddleware, logger=async_logger)
    ]
)
```

My project uses `mypy` to test typing, and this passed using `0.32.0`

https://github.com/encode/starlette/pull/2381 changed `starlette.middleware.Middleware`'s first argument from type `Any` to `type[_MiddlewareClass[P]]`

I upgraded to `mypy` `1.10.0` and `starlette` `0.37.2`.

Now when I run `mypy`  I get an error like
```
error: Argument 1 to "Middleware" has incompatible type "type[TraceTokenLoggerMiddleware]"; expected "type[_MiddlewareClass[[AsyncLogger]]]"  [arg-type]
```

Please advise how to stop this error or if a fix is needed in `starlette`.

It seems like I ought to make `MyMiddleware` inherit from `_MiddlewareClass`?! But it's private so I shouldn't do that!

**Comment by ShedPlant** [comment]
FWIW if I do define `class TraceTokenLoggerMiddleware(_MiddlewareClass):`  and change the type hints of `__call__`, then `mypy` stops complaining.

So I suggest `_MiddlewareClass` be made public, renamed to `MiddlewareAbstractClass` or something.

**Reply by pawelrubin** [reply]
The `_MiddlewareClass` is a `Protocol` meant for _structural subtyping_, not for _subclassing_. Check out [this SO answer](https://stackoverflow.com/a/73070133/8091093) for a neat explanation.

I see the point in making the type "public" though. Perhaps we could add it to `starlette.types`? WDYT @adriangb?

**Reply by adriangb** [reply]
I don’t think that’s really necessary. You can always import it as private, define the class, fix any type errors then remove the import / subclass.

**Comment by ShedPlant** [comment]
@pawelrubin and @adriangb , are you able to comment / help please?

**Comment by pawelrubin** [comment, marked as answer]
You should be able to fix the issue by providing complete type annotations to your `TraceTokenLoggerMiddleware`:

```python
from typing import Any

from starlette.middleware import Middleware
from starlette.types import ASGIApp, Receive, Scope, Send


IAsyncLogger = Any  # whatever the type is


class TraceTokenLoggerMiddleware:
    def __init__(self, app: ASGIApp, logger: IAsyncLogger) -> None:
        self._app = app
        self._async_logger = logger

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # do something
        pass

async_logger: IAsyncLogger = NotImplemented

Middleware(TraceTokenLoggerMiddleware, logger=async_logger)  # no errors here
```

**Reply by ShedPlant** [reply]
Thanks, I can confirm `mypy` doesn't complain anymore if I do that.

Although, my IDE (PyCharm Community Edition) does

![image](https://github.com/encode/starlette/assets/13786582/af430995-f9e9-41ea-833c-a56e97f5f3e4)

"Expected type 'Type[_MiddlewareClass[ParamSpec("P")]]', got 'Type[TraceTokenLoggerMiddleware]' instead
